rust/src/storage/s3/dynamodb_lock.rs

```diff
@@ -0,0 +1,510 @@
+    /// The amount of time to wait before trying to get the lock again.
+    pub refresh_period: Duration,
+    /// The amount of time to wait in addition to `lease_duration`.
+    pub additional_time_to_wait_for_lock: Duration,
+}
+
+impl Default for Options {
+    fn default() -> Self {
+        fn str_env(key: &str, default: String) -> String {
+            std::env::var(key).unwrap_or(default)
+        }
+
+        fn u64_env(key: &str, default: u64) -> u64 {
+            std::env::var(key)
+                .ok()
+                .and_then(|e| e.parse::<u64>().ok())
+                .unwrap_or(default)
+        }
+
+        let refresh_period = Duration::from_secs(u64_env(options::REFRESH_PERIOD_SECS, 1000));
+        let additional_time_to_wait_for_lock =
+            Duration::from_secs(u64_env(options::ADDITIONAL_TIME_TO_WAIT_SECS, 1000));
+
+        Self {
+            partition_key_value: str_env(options::PARTITION_KEY_VALUE, "delta-rs".to_string()),
+            table_name: str_env(options::TABLE_NAME, "delta_rs_lock_table".to_string()),
+            owner_name: str_env(options::OWNER_NAME, Uuid::new_v4().to_string()),
+            lease_duration: u64_env(options::LEASE_DURATION, 20),
+            refresh_period,
+            additional_time_to_wait_for_lock,
+        }
+    }
+}
+
+/// A lock that has been successfully acquired
+#[derive(Clone, Debug)]
+pub struct LockItem {
+    /// The name of the owner that owns this lock.
+    pub owner_name: String,
+    /// Current version number of the lock in DynamoDB. This is what tells the lock client
+    /// when the lock is stale.
+    pub record_version_number: String,
+    /// The amount of time (in seconds) that the owner has this lock for.
+    pub lease_duration: u64,
+    /// Tells whether or not the lock was marked as released when loaded from DynamoDB.
+    pub is_released: bool,
+    /// Optional data associated with this lock.
+    pub data: Option<String>,
+    /// The last time this lock was updated or retrieved.
+    pub lookup_time: u128,
+}
+
+impl LockItem {
+    fn is_expired(&self) -> bool {
+        if self.is_released {
+            return true;
+        }
+        now_millis() - self.lookup_time > (self.lease_duration as u128) * 1000
+    }
+}
+
+/// Error returned by the [`DynamoDbLockClient`] API.
+#[derive(thiserror::Error, Debug)]
+pub enum DynamoError {
+    /// Error caused by the DynamoDB table not being created.
+    #[error("Dynamo table not found")]
+    TableNotFound,
+
+    /// Error that indicates the condition in the DynamoDB operation could not be evaluated.
+    /// Mostly used by [`DynamoDbLockClient::acquire_lock`] to handle unsuccessful retries
+    /// of acquiring the lock.
+    #[error("Conditional check failed")]
+    ConditionalCheckFailed,
+
+    /// Error that is returned by [`DynamoDbLockClient::release_lock`] when the given lock
+    /// has been already expired and could not be released.
+    #[error("Lock is expired")]
+    LockIsExpired,
```

Comment:
Is this naming ok?

Reply:
yep, can probably be further simplified to `LockExpired`.